### PR TITLE
DBZ-1052 Transaction metadata

### PR DESCRIPTION
**Pull request by jpechane**
https://issues.redhat.com/browse/DBZ-1052

### CODE REVIEW

**Comment by jpechane** [comment]
Depends on https://github.com/debezium/postgres-decoderbufs/pull/24

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java [hunk 1]
```diff
@@ -379,6 +379,70 @@ public void update() throws Exception {
...
+        connection.execute(inserts);
+        connection.setAutoCommit(true);
+
+        connection.execute("INSERT INTO tableb VALUES(1000, 'b')");
+
+        // BEGIN, data, END, BEGIN, data
+        final SourceRecords records = consumeRecordsByTopic(1 + RECORDS_PER_TABLE * 2 + 1 + 1 + 1);
+        final List<SourceRecord> tableA = records.recordsForTopic("server1.dbo.tablea");
+        final List<SourceRecord> tableB = records.recordsForTopic("server1.dbo.tableb");
+        final List<SourceRecord> tx = records.recordsForTopic("__debezium.transaction.server1");
```
Comment (gunnarmorling) [inline debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java]: As that's a user-facing topic -- unlike heartbeats -- I'd suggest to stick to the established naming pattern for topics and make it `<logical name>.transactions`, e.g. `dbserver1.transactions`.

File: debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java [hunk 2]
```diff
@@ -431,6 +431,9 @@ else if (schema.type() == Type.INT8 || schema.type() == Type.INT16 || schema.typ
         else if (schema.type() == Type.STRING || schema.type() == Type.INT64) {
             ceExtensionSchema = SchemaBuilder.string();
         }
+        else if (schema.type() == Type.STRUCT) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java]: Have you tested this? I'd expect that the nested attributes will need proper handling, i.e. prefixing them with with some constant literal as they are all put into one flat namespace. That said I'm a bit dubious about the nested structure begin with. This ties into my question about the purpose of these attributes in change events themselves. I kinda thought we'd _only_ need the TX id, in which case the question of nesting is a moot point.

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java [hunk 3]
```diff
@@ -69,4 +70,6 @@
      * Records the name of the collection and the timestamp of the last event
      */
     void event(DataCollectionId collectionId, Instant timestamp);
+
+    TransactionContext getTransactionContext();
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: Please add a comment, as it's a public contract and it'll be beneficial for us, too.

**Comment by jpechane** [comment]
> I don't see TX metadata records during snapshotting; is this intended, do you plan to add it later on?

It is intended, this will be more complicated to iunify accross connectors so I'd add it as follow-up

> I tried the PG tutorial example and I observed an unexpected behaviour: the END event will only be sent once another TX has begun; is this because we don't receive a commit event (I'm using decoderbufs, do I need to update decoderbufs in the DB)?

Exactly tha't why the is the dcoderbufs change, pgoutput should work OOTB

> The BEGIN event shouldn't have the "event_count": null block, i.e. likely a separate schema is warranted.

Yes but that would require specific configuration of schema registry so I opted fo single schema per topic

> For the END event I was assuming there'd be also the count-per-table. Not strictly needed for an initial cut but we might consider it for a later round.

Right, this really should not be difficult to add. I was unsure whether to add it or not so easy to add as a follow-up

> What's the purpose / use case for "total_order": 1 and "data_collection_order": 1 in change events themselves? Can we get away with just the TX id in CDC events?

This would help downstream client to identify which event it is missing and deduplicate in case of duplicates

File: debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java]: Yes, CE test fails due to this reason, thus I added this as a hotfix - for whether we need or not other information please see other comment.

File: debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/TransactionMetadataIT.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    private static final String INSERT_STMT = "INSERT INTO s1.a (aa) VALUES (1);" +
+            "INSERT INTO s2.a (aa) VALUES (1);";
+    private static final String SETUP_TABLES_STMT = "DROP SCHEMA IF EXISTS s1 CASCADE;" +
+            "DROP SCHEMA IF EXISTS s2 CASCADE;" +
+            "CREATE SCHEMA s1; " +
+            "CREATE SCHEMA s2; " +
+            "CREATE TABLE s1.a (pk SERIAL, aa integer, PRIMARY KEY(pk));" +
+            "CREATE TABLE s2.a (pk SERIAL, aa integer, bb varchar(20), PRIMARY KEY(pk));" +
+            INSERT_STMT;
+    private PostgresConnector connector;
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-postgres/src/test/java/io/debezium/connector/postgresql/TransactionMetadataIT.java]: Removed

File: debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java (hunk 1 above)
Comment (jpechane) [reply debezium-connector-sqlserver/src/test/java/io/debezium/connector/sqlserver/SqlServerConnectorIT.java]: Done

File: debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java (hunk 3 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/pipeline/spi/OffsetContext.java]: Added

**Comment by gunnarmorling** [comment]
> It is intended, this will be more complicated to iunify accross connectors so I'd add it as follow-up

Ok; can you log a follow-up JIRA so we can begin capturing these things?

> but that would require specific configuration of schema registry

Ok, fair enough.

>  was unsure whether to add it or not so easy to add as a follow-up

+1 Add it to that new JIRA issue then, please.

> This would help downstream client...

Ok, but it seems we would get away with the single total events in TX counter then? Also keep in mind it's not the use case we're aiming for (which would be aggregate materialization), so I'd remove at least one if not both of them for now.

Let's talk about the nesting/flat represenation in the meeting later today.

**Comment by gunnarmorling** [comment]
@jpechane, something that just crossed my mind: if you're strongly in favour of the nested structure, how about having it _next_ to `before`, `after`,`source` etc. instead of _within_ `source`:

```
{
  "before": {
    "id": 1003,
    "first_name": "Edward",
    "last_name": "Walker",
    "email": "ed@walker.com"
  },
  "after": {
    "id": 1003,
    "first_name": "Kobe",
    "last_name": "Walker",
    "email": "ed@walker.com"
  },
  "source": {
    "version": "1.1.0-SNAPSHOT",
    "connector": "postgresql",
    "name": "dbserver1",
    "ts_ms": 1580214030129,
    "snapshot": "false",
    "db": "postgres",
    "schema": "inventory",
    "table": "customers",
    "txId": 604,
    "lsn": 34193312,
    "xmin": null
  },
  "transaction": {
    "id": "604",
    "total_order": 1,
    "data_collection_order": 1
  },
  "op": "u",
  "ts_ms": 1580214030134
}
```

File: debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java (hunk 2 above)
Comment (jpechane) [reply debezium-core/src/main/java/io/debezium/converters/CloudEventsConverter.java]: Removed as transaction is no longer in source info block

**Comment by jpechane** [comment]
@gunnarmorling  There is only one follow-up task left - https://issues.redhat.com/browse/DBZ-1742
counts-per-table are now also stored in `END` tx event
And tx is now stored in a separate block

**Comment by gunnarmorling** [comment]
@jpechane, could you also add the new attributes (as a flattened set of attributes) to the CE converter? And there should be mention in the documentation, too.

**Comment by jpechane** [comment]
@gunnarmorling We've hit a problem here. `data_collections` is an array of structs which does not have a counterpart in CloudEvents. So recommend to encode it as a string in format `['table1':3, 'table2':5]` so more or less as a JSON array string. Any objections?

**Comment by gunnarmorling** [comment]
> data_collections is an array of structs which does not have a counterpart in CloudEvents. 

I can't find `data_collections` in this PR, is this for the follow-up already?

**Comment by jpechane** [comment]
@gunnarmorling CE support (only for data messages) is in

**Comment by gunnarmorling** [comment]
Here's an idea for future evolution that just came to my mind: now that we have the TX topic, we may have support in the connector for enriching these events with application-provided metadata as discussed in the [auditing blog post](https://debezium.io/blog/2019/10/01/audit-logs-with-change-data-capture-and-stream-processing/). For a configured table, we wouldn't emit its INSERTs as regular CDC events, but add that info into the corresponding record of the TX topic.

**Comment by gunnarmorling** [comment]
Is there a follow-up DBZ issue already where we could track this idea?

**Comment by jpechane** [comment]
@gunnarmorling Should we call it transaction topic then?

**Comment by gunnarmorling** [comment]
Isn't it called that already?

>

**Comment by jpechane** [comment]
@gunnarmorling I mean, maybe the name should be different if we use it for other purposes as well

**Comment by jpechane** [comment]
@gunnarmorling - https://issues.redhat.com/browse/DBZ-1456 - I've added it to 1.1-backlog

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java [hunk 5]
```diff
@@ -130,13 +131,6 @@ public boolean shouldMessageBeSkipped(ByteBuffer buffer, Long lastReceivedLsn, L
                     // same session a `R` message was followed by an insert/update/delete message.
                     // @formatter:on
                 case COMMIT:
-                    // For now skip these message types so that the LSN associated with the message won't
-                    // be flushed back to PostgreSQL. There is a potential LSN assignment concern with
-                    // ReplicationConnectionIT#testHowRelationMessagesAreReceived where the first COMMIT
```
[the remark this answers is not in the document]
Comment (jpechane) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java]: @gunnarmorling No, I've not encountered yet. I've discussed the matter with @Naros and we came to conclusion that if the problem is no longer viisble in tests that we can remove it.

**Comment by jpechane** [comment]
@gunnarmorling The config options definitions were added to connectors and I finally lost patience - https://issues.redhat.com/browse/DBZ-1750

is there anything left but the docuemntation?

**Comment by jpechane** [comment]
@gunnarmorling I've added docs for CE and postgres. When reviewd and accepted I'll copy the text to sqlserver and Oracle too.

File: documentation/modules/ROOT/pages/integrations/cloudevents.adoc [hunk 6]
```diff
@@ -97,6 +101,9 @@ Alternatively, Avro can be used as content type for the `data` attribute:
   "iodebeziumtxId" : "578",
   "iodebeziumlsn" : "33227720",
   "iodebeziumxmin" : null,
+  "iodebeziumtxid": "589",
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/integrations/cloudevents.adoc]: There is `iodebeziumtxId` already (see above). That's really what I'm a bit concerned about -- the redundancy of this new info. I don't have a good answer right now. Maybe we should deprecate and eventually remove the current txId attribute in the source block?

Comment (jpechane) [reply documentation/modules/ROOT/pages/integrations/cloudevents.adoc]: @gunnarmorling I don't think that it is necessary to remove the redundancy. Don't forget that transaction metadata are optional. So I'd keep it in source info.
In fact this looks like a problem only in flattened format and in this case I propose to just document it and/or change `iodebeziumtx` to a different one so the difference is more visible, maybe `iodebeziumTxn`?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java [hunk 7]
```diff
@@ -126,15 +128,33 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
                         dispatcher.dispatchHeartbeatEvent(offsetContext);
                         return;
                     }
+                    if (TransactionMessage.isTransactionalMessage(message) && !connectorConfig.shouldProvideTransactionMetadata()) {
+                        LOGGER.trace("Received transactional message {}", message);
+                        return;
```
Comment (gunnarmorling) [inline debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/PostgresStreamingChangeEventSource.java]: Isn't the early return a problem as we'd miss the update of `lastCompletelyProcessedLsn` below?

File: debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply debezium-connector-postgres/src/main/java/io/debezium/connector/postgresql/connection/pgoutput/PgOutputMessageDecoder.java]: Ok, thanks!

File: documentation/modules/ROOT/pages/integrations/cloudevents.adoc [hunk 8]
```diff
@@ -97,6 +101,9 @@ Alternatively, Avro can be used as content type for the `data` attribute:
   "iodebeziumtxId" : "578",
   "iodebeziumlsn" : "33227720",
   "iodebeziumxmin" : null,
+  "iodebeziumtxid": "578",
```
Comment (gunnarmorling) [inline documentation/modules/ROOT/pages/integrations/cloudevents.adoc]: Ok, so I'm still not happy about this; but no blocker for now. I've added it to [DBZ-1701](https://issues.redhat.com/browse/DBZ-1701) to get rid of `iodebeziumtxId` if `iodebeziumtxid` is present.

**Comment by gunnarmorling** [comment]
@jpechane, there's an NPE in one of the tests now :(

**Comment by jpechane** [comment]
@gunnarmorling Fixed another test - looks like a timing issue cause by resent changes in `wal2json` empty message processing

**Comment by gunnarmorling** [comment]
🎉  Woot, it's merged. Excellent, thanks a ton, @jpechane!
